modules/auxiliary/spoof/nbns/nbns_response.rb

```diff
@@ -44,108 +47,144 @@ def initialize
     ])
 
     register_advanced_options([
-      OptBool.new('Debug', [ false, "Determines whether incoming packet parsing is displayed", false])
+      OptBool.new('DEBUG', [ false, "Determines whether incoming packet parsing is displayed", false])
     ])
 
     deregister_options('RHOST', 'PCAPFILE', 'SNAPLEN', 'FILTER')
+    self.thread = nil
+    self.sock = nil
   end
 
-  def run
-    check_pcaprub_loaded() # Check first since otherwise this is all for naught
-    # MacOS X workaround
-    ::Socket.do_not_reverse_lookup = true
+  def dispatch_request(packet, rhost, src_port)
+    rhost = ::IPAddr.new(rhost)
+     # `recvfrom` (on Linux at least) will give us an ipv6/ipv4 mapped
+    # addr like "::ffff:192.168.0.1" when the interface we're listening
+    # on has an IPv6 address. Convert it to just the v4 addr
+    if rhost.ipv4_mapped?
+      rhost = rhost.native
+    end
 
-    @sock = ::UDPSocket.new()
-    @sock.setsockopt(::Socket::SOL_SOCKET, ::Socket::SO_REUSEADDR, 1)
-    @sock.bind('', 137) # couldn't specify srv host because it missed broadcasts
+    # Convert to string
+    rhost = rhost.to_s
 
-    @run = true
+    spoof = ::IPAddr.new(datastore['SPOOFIP'])
 
-    print_status("NBNS Spoofer started. Listening for NBNS requests...")
+    return if packet.length == 0
 
-    begin
+    nbnsq_transid      = packet[0..1]
+    nbnsq_flags        = packet[2..3]
+    nbnsq_questions    = packet[4..5]
+    nbnsq_answerrr     = packet[6..7]
+    nbnsq_authorityrr  = packet[8..9]
+    nbnsq_additionalrr = packet[10..11]
+    nbnsq_name         = packet[12..45]
+    decoded = ""
+    nbnsq_name.slice(1..-2).each_byte do |c|
+      decoded << "#{(c - 65).to_s(16)}"
+    end
+    nbnsq_decodedname = "#{[decoded].pack('H*')}".strip()
+    nbnsq_type         = packet[46..47]
+    nbnsq_class        = packet[48..49]
+
+    return unless nbnsq_decodedname =~ /#{datastore['REGEX']}/i
```

Comment:
Minor, especially because this is just reogranized code, but I'd do this differently.  I'd compile `datastore['REGEX']` into `@regex` in `setup` and avoid the constant recompilation of the regex.